### PR TITLE
New package: NoiseTorch-0.11.4

### LINKED ISSUE #27677: package request https://github.com/lawl/NoiseTorch
### System

* xuname:  
  *output of ``xuname`` (part of xtools)*
* package:  
  *affected package(s) including the version*: ``xbps-query -p pkgver <pkgname>``

### Steps to reproduce the behavior

https://github.com/lawl/NoiseTorch
this is really handy and i think it would be very nice to have in void linux, its bascially https://krisp.ai/ but for linux and free and open source :D

### PR DESCRIPTION
**Pull request by uncomfyhalomacro**
#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

#### Does it build and run successfully? 
(Please choose at least one native build and, if supported, at least one cross build. More are better.)
- [x] I built this PR locally for my native architecture, (ARCH-LIBC)
- [x] I built this PR locally for these architectures (if supported. mark crossbuilds):
  - [x] x86_64-musl
  - [x] i686
  - [x] aarch64
  - [x] aarch64-musl
  - [x] armv7l
  - [x] armv6l-musl

Closes #27677

### CODE REVIEW

File: srcpkgs/noisetorch/template [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+archs="x86_64*"
+wrksrc=NoiseTorch-${version}
+build_style=gnu-makefile
+hostmakedepends="go"
+depends="polkit"
+short_desc="Real-time microphone noise suppression on Linux"
+maintainer="Soc Virnyl S. Estela <renegan.ronin@gmail.com>"
+license="GPL-3.0-or-later"
+homepage="https://github.com/lawl/NoiseTorch"
+distfiles="https://github.com/lawl/NoiseTorch/archive/${version}.tar.gz"
```
Comment (kawaiiamber) [inline srcpkgs/noisetorch/template:14]: Since the homepage is contained within the distfiles url, common for github projects, a cleaner expression that is often used in VOID packages: `distfiles="${homepage}/archive/${version}.tar.gz"`

File: srcpkgs/noisetorch/template [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Template file for 'noisetorch'
+pkgname=noisetorch
+version=0.11.3
+revision=1
+archs="x86_64*"
```
Comment (kawaiiamber) [inline srcpkgs/noisetorch/template:5]: Even though this will include musl archs, why limit archs at all? Does the build fail otherwise? If so, is it fixable? It's much preferable to fix cross build than give up on it.

File: srcpkgs/noisetorch/template [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+wrksrc=NoiseTorch-${version}
+build_style=gnu-makefile
+hostmakedepends="go"
+depends="polkit"
+short_desc="Real-time microphone noise suppression on Linux"
+maintainer="Soc Virnyl S. Estela <renegan.ronin@gmail.com>"
+license="GPL-3.0-or-later"
```
Comment (kawaiiamber) [inline srcpkgs/noisetorch/template:12]: While the project is release under GPL version 3, it's not clear whether they intend for it to be released under later versions of the license as well. Usually that is included in the README or source files. I couldn't find such labels, however. Furthermore, the `untar.go` file included might be licensed differently. From `untar.go`:
```
/*Copyright (c) 2009 The Go Authors. All rights reserved.Redistribution and use in source and binary forms, with or withoutmodification, are permitted provided that the following conditions aremet:   * Redistributions of source code must retain the above copyrightnotice, this list of conditions and the following disclaimer.   * Redistributions in binary form must reproduce the abovecopyright notice, this list of conditions and the following disclaimerin the documentation and/or other materials provided with thedistribution.   * Neither the name of Google Inc. nor the names of itscontributors may be used to endorse or promote products derived fromthis software without specific prior written permission.THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOTLIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FORA PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHTOWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOTLIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANYTHEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USEOF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.*/
// This is yoinked from https://github.com/golang/build/blob/master/internal/untar/untar.go// which is unfortunately an internal package we can't import, but it's exactly what we need.// So just copy paste it.
```
They just... "copy paste it." Seems odd.

**Review by kawaiiamber — COMMENTED** [review]
Curious about fixing cross build if it's not working.

**Comment by kawaiiamber** [comment]
Your commit message is currently "Add new package noisetorch", you should rebase and change the commit message to "New package: noisetorch-0.11.3", use a dash after pkgname then version. PR name could also replace that space with a dash.

File: srcpkgs/noisetorch/template (hunk 1 above)
Comment (paper42) [reply srcpkgs/noisetorch/template:14]: @kawaiiamber this used to be the case, but templates are moving away from this style.

File: srcpkgs/noisetorch/template (hunk 3 above)
Comment (ReneganRonin) [reply srcpkgs/noisetorch/template:12]: I guess I should ask the project maintainers about this. Seems odd indeed.

File: srcpkgs/noisetorch/template (hunk 2 above)
Comment (ReneganRonin) [reply srcpkgs/noisetorch/template:5]: I will try to build it through musl later. but I cannot confirm that it builds on other architectures. How should I go about this?

File: srcpkgs/noisetorch/template (hunk 1 above)
Comment (ReneganRonin) [reply srcpkgs/noisetorch/template:14]: Seems reasonable and cleaner for me to use `distfiles="${homepage}/archive/${version}.tar.gz"`

Comment (paper42) [reply srcpkgs/noisetorch/template]: see #27735

File: srcpkgs/noisetorch/template (hunk 2 above)
Comment (paper42) [reply srcpkgs/noisetorch/template:5]: Try building it in a musl masterdir or cross compile for example to aarch64-musl. Of course you will have to remove this line.

File: srcpkgs/noisetorch/template (hunk 3 above)
Comment (paper42) [reply srcpkgs/noisetorch/template:12]: That is a BSD 3 Clause license, I am pretty sure you don't have to care about it.

File: srcpkgs/noisetorch/template (hunk 1 above)
Comment (ReneganRonin) [reply srcpkgs/noisetorch/template]: @paper42 it doesnt seem like they reached to a conclusion. now i am confused...

**Comment by ReneganRonin** [comment]
@kawaiiamber @paper42 seems like x86_64-musl failed to build because polkit failed to build. 🤔

File: srcpkgs/noisetorch/template (hunk 1 above)
Comment (kawaiiamber) [reply srcpkgs/noisetorch/template]: Correct, the discussion is currently inconclusive - feel free to use either, just thought I'd mention since I was reviewing, anyway.

File: srcpkgs/noisetorch/template (hunk 3 above)
Comment (kawaiiamber) [reply srcpkgs/noisetorch/template:12]: You're right - still not sure if GPL-3.0-only or GPL-3.0-or-later

**Comment by kawaiiamber** [comment]
> @kawaiiamber @paper42 seems like x86_64-musl failed to build because polkit failed to build. 🤔 

Hmmmm, I can try building with the template you wrote when I get return home. What happens if you try just building polkit?

**Comment by ReneganRonin** [comment]
@kawaiiamber Tests show 4 fails in `make` (i saw some segfaults). I am retrying right now. Idk if thats because of my CPU. I have a very old laptop after all. 🥲

**Comment by paper42** [comment]
I think you might be doing something wrong. On your main PC, try running these commands, polkit shouldn't need to be built.
```
git pull --autostash --rebase upstream master
./xbps-src -m masterdir-musl binary-bootstrap x86_64-musl
./xbps-src -m masterdir-musl pkg noisetorch
```

**Comment by kawaiiamber** [comment]
> @kawaiiamber Tests show 4 fails in `make` (i saw some segfaults). I am retrying right now. Idk if thats because of my CPU. I have a very old laptop after all. 🥲

I assume these segfaults only appear in the musl arch? CPU won't result in segmentation faulting here; it's not because of your CPU. I'll test when I return home. It's possible musl arch might need patches applied, who knows. I wanna use your template myself so I can see what error specifically happened durring the make command.

File: srcpkgs/noisetorch/template (hunk 1 above)
Comment (ericonr) [reply srcpkgs/noisetorch/template]: It's not long enough to require using `homepage`, but it's fine as is now as well.

File: srcpkgs/noisetorch/template (hunk 2 above)
Comment (ericonr) [reply srcpkgs/noisetorch/template:5]: Please remove the restriction so we can see build results on CI.

File: srcpkgs/noisetorch/template (hunk 3 above)
Comment (ericonr) [reply srcpkgs/noisetorch/template:12]: Ugh it has an additional clause on top of GLP 3: https://github.com/lawl/NoiseTorch/blob/master/LICENSE#L5

Using https://github.com/lawl/NoiseTorch/blob/master/c/ladspa/module.c as the most restrictive file, the final package should be GPL-3.0-or-later. No idea how to mention the additional clause.,

**Comment by ReneganRonin** [comment]
> I think you might be doing something wrong. On your main PC, try running these commands, polkit shouldn't need to be built.
> 
> ```
> git pull --autostash --rebase upstream master
> ./xbps-src -m masterdir-musl binary-bootstrap x86_64-musl
> ./xbps-src -m masterdir-musl pkg noisetorch
> ```

It succesfully built. going to check if it runs

File: srcpkgs/noisetorch/template (hunk 3 above)
Comment (paper42) [reply srcpkgs/noisetorch/template:12]: Rust has a similar restriction (not mentioned in the license) and we don't mention it in the template.

**Comment by paper42** [comment]
@ReneganRonin force push the rebased version to this PR, that will make the CI run

**Comment by ReneganRonin** [comment]
@paper42 it is now ready. :D thanks all of you. i hope this PR gets merged

**Comment by ReneganRonin** [comment]
@paper42 @kawaiiamber @ericonr seems like it only builds for archs i686, x86_64, and x86_64-musl. Should I specify such archs in the template, commit and rebase, and push the changes now?

**Comment by kawaiiamber** [comment]
> @paper42 @kawaiiamber @ericonr seems like it only builds for archs i686, x86_64, and x86_64-musl. Should I specify such archs in the template, commit and rebase, and push the changes now?

No need to rebase - just git commit --amend

**Comment by ReneganRonin** [comment]
@kawaiiamber okay already pushed the changes. waiting for approval. thanks :D

**Comment by paper42** [comment]
It builds for other architectures too, but cross compiling is broken. The upstream build system is not great.

**Comment by ReneganRonin** [comment]
> It builds for other architectures too, but cross compiling is broken. The upstream build system is not great.

i guess i should remove that line again. how did u know it build for other architectures? i am not very familiar with this stuff

**Comment by ReneganRonin** [comment]
@kawaiiamber @paper42 @ericonr Is this PR ready to be merged?

**Comment by ericonr** [comment]
Try using the Go build style instead, it might just work.

**Comment by paper42** [comment]
It will not, it has a C part and it needs to build an executable (from go sources) for the host machine to embed a license to the source. After that it's a traditional go build.

**Comment by ericonr** [comment]
@paper42 I'd consider a `pre_build` function for the `c/ladspa` directory, then.

**Comment by paper42** [comment]
A template like this works, but I am not sure if it's pretty enough.
```
# Template file for 'noisetorch'
pkgname=noisetorch
version=0.11.3
revision=1
wrksrc=NoiseTorch-${version}
build_style=go
go_ldflags="-X main.version=${version}"
go_import_path="."
depends="polkit"
short_desc="Real-time microphone noise suppression on Linux"
maintainer="Soc Virnyl S. Estela <renegan.ronin@gmail.com>"
license="GPL-3.0-or-later"
homepage="https://github.com/lawl/NoiseTorch"
distfiles="${homepage}/archive/${version}.tar.gz"
checksum=b67155211278affc72f15d0ded9da560524e7bb0659012f9d41e69b531b9265c

pre_build() {
	make rnnoise
	GOARCH= go build scripts/embedlicenses.go
	./embedlicenses
}

post_install() {
	vinstall assets/noisetorch.desktop 644 usr/share/applications
	vinstall assets/icon/noisetorch.png 644 usr/share/icons/hicolor/256x256/apps
}
```

**Comment by paper42** [comment]
I forgot to set the distribution variable to "void" in the same place where version is set.

**Comment by ReneganRonin** [comment]
thats fine. im going to use that template tomorrow. just got busy on personal stuff lately. thanks. :D

**Comment by kawaiiamber** [comment]
There's currently three commits - be sure to rebase to one. Just use `git commit --amend` for future commits.

**Comment by ReneganRonin** [comment]
@kawaiiamber @ericonr is this ready to be merged now? or nay?

**Comment by kawaiiamber** [comment]
> @kawaiiamber @ericonr is this ready to be merged now? or nay?

I'd go test it out. `xbps-install --repository=/path/to/package.xbps` and see if it runs correctly.

**Comment by ReneganRonin** [comment]
it works for me :D

**Comment by kawaiiamber** [comment]
> it works for me :D

Still squash those two commits to one: `git rebase -i HEAD~2` and squash the last one.

**Comment by ReneganRonin** [comment]
> > it works for me :D
> 
> Still squash those two commits to one: `git rebase -i HEAD~2` and squash the last one.

okay it is ready now. i nearly messed up lol.

File: srcpkgs/noisetorch/template [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+pre_build() {
+	make rnnoise
+	GOARCH= go build scripts/embedlicenses.go
+	./embedlicenses
+}
+
+post_install() {
+	# Fixes icon because it does not show
+	sed 's/Icon=noisetorch.png/Icon=noisetorch/' assets/noisetorch.desktop > noisetorch.desktop
```
Comment (paper42) [inline srcpkgs/noisetorch/template:26]: why is this line necessary?

File: srcpkgs/noisetorch/template [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+go_ldflags="-X main.version=${version} -X main.distribution=voidlinux"
+go_import_path="."
+depends="polkit"
+short_desc="Real-time microphone noise suppression on Linux"
+maintainer="Soc Virnyl S. Estela <renegan.ronin@gmail.com>"
+license="GPL-3.0-or-later"
+homepage="https://github.com/lawl/NoiseTorch"
+distfiles="${homepage}/archive/${version}.tar.gz"
+checksum=b67155211278affc72f15d0ded9da560524e7bb0659012f9d41e69b531b9265c
+nopie=yes
```
Comment (paper42) [inline srcpkgs/noisetorch/template:16]: why? can you leave a comment?

File: srcpkgs/noisetorch/template (hunk 4 above)
Comment (ReneganRonin) [reply srcpkgs/noisetorch/template:26]: if left as is, the desktop icon does not show. but weirdly enough, that fixes it...

File: srcpkgs/noisetorch/template (hunk 5 above)
Comment (ReneganRonin) [reply srcpkgs/noisetorch/template:16]: im going to delete that line. that was from the old template i have before.

**Comment by ReneganRonin** [comment]
@paper42 

Leaving the desktop file as is:
screenshot-0624161754

After fixing the icon in the desktop file:
image

**Comment by paper42** [comment]
This issue seems to happen only with some implementations, so I think patching it in the template is fine. Could you make a PR upstream?

File: srcpkgs/noisetorch/template [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# Template file for 'noisetorch'
+pkgname=noisetorch
+version=0.11.3
+revision=1
+wrksrc=NoiseTorch-${version}
+build_style=go
+go_ldflags="-X main.version=${version} -X main.distribution=voidlinux"
+go_import_path="."
+depends="polkit"
```
Comment (lawl) [inline srcpkgs/noisetorch/template:9]: Hello, upstream maintainer here. @ReneganRonin linked this PR upstream and i look the liberty to quickly glance over it.

Since you set `CAP_SYS_RESOURCE` in the package and the binary is managed by your package manager, polkit should not be a dependency. We only use `pkexec` from polkit if the binary manages itself. As long as the binary always has `CAP_SYS_RESOURCE` this should not be required.

Comment (ReneganRonin) [reply srcpkgs/noisetorch/template:9]: I see im going to remove that dependency then. thanks for the suggestions :D

**Comment by ReneganRonin** [comment]
### Summary of changes:
- Fix build style from `gnu_makefile` to `go` by adding a `pre_build()` function.
- Removed `nopie` unnecessary line.
- Added patch `desktop-icon.patch` and `series` file to `patches` folder to fix desktop icon issue.
- Removed `polkit` as dependency since upstream maintainer has suggested it is not required since I set `CAP_SYS_RESOURCE`.

@ericonr @paper42 @kawaiiamber thanks for the assistance. Any more you can suggest for the template? \:D

**Comment by paper42** [comment]
@ReneganRonin the sed line was better than a whole patch, just use vsed instead.

**Comment by paper42** [comment]
lgtm :+1:

**Comment by kawaiiamber** [comment]
> ### Summary of changes:
> * Fix build style from `gnu_makefile` to `go` by adding a `pre_build()` function.
> * Removed `nopie` unnecessary line.
> * ~Added patch `desktop-icon.patch` and `series` file to `patches` folder to fix desktop icon issue.~
> * Fixed desktop icon issue using `vsed`.
> * Removed `polkit` as dependency since upstream maintainer has suggested it is not required since I set `CAP_SYS_RESOURCE`.
> 
> @ericonr @paper42 @kawaiiamber thanks for the assistance. Any more you can suggest for the template? :D

LGTM

**Comment by kawaiiamber** [comment]
Since upstream has fixed the Icon spec by ommiting `.png` via [#156](https://github.com/lawl/NoiseTorch/pull/156) into master, I wonder if a new release tag will be made - sometimes projects make release tags for bug fixes. If that's the case, it might be worth waiting to get rid of the `vsed` step in `post_install`. That's just my two cents.

**Comment by ReneganRonin** [comment]
> Since upstream has fixed the Icon spec by ommiting `.png` via [#156](https://github.com/lawl/NoiseTorch/pull/156) into master, I wonder if a new release tag will be made - sometimes projects make release tags for bug fixes. If that's the case, it might be worth waiting to get rid of the `vsed` step in `post_install`. That's just my two cents.

for me, i think adding the `vsed` is fine since it fixes the Icon spec problem. Idk if other contributors agree but if the majority tells me to remove `vsed` then I will remove it. i think users can fix it by themselves until there is a new release. it is not that hard.

**Comment by kawaiiamber** [comment]
> > Since upstream has fixed the Icon spec by ommiting `.png` via [#156](https://github.com/lawl/NoiseTorch/pull/156) into master, I wonder if a new release tag will be made - sometimes projects make release tags for bug fixes. If that's the case, it might be worth waiting to get rid of the `vsed` step in `post_install`. That's just my two cents.
> 
> for me, i think adding the `vsed` is fine since it fixes the Icon spec problem. Idk if other contributors agree but if the majority tells me to remove `vsed` then I will remove it. i think users can fix it by themselves until there is a new release. it is not that hard.

????????????
I'm not suggesting to remove it - I'm saying another release tag might be released that has this patch already built in and you can update PR to that - release tags for upstream seem to happen about every 30 days - ish.

**Comment by ReneganRonin** [comment]
oh okay. sorry i misinterpreted what u said :grimacing:

**Comment by paper42** [comment]
imo it's not worth waiting for a new release just to remove one line. On the next update, vsed will warn that the line doesn't do anything.

**Comment by ReneganRonin** [comment]
> 
> 
> 
> 
> #### General
> 
> - [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)
> 
> 
> 
> #### Have the results of the proposed changes been tested?
> 
> - [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
> 
> - [ ] I generally don't use the affected packages but briefly tested this PR
> 
> 
> 
> 
> 
> 
> 
> #### Does it build and run successfully? 
> 
> (Please choose at least one native build and, if supported, at least one cross build. More are better.)
> 
> - [x] I built this PR locally for my native architecture, (ARCH-LIBC)
> 
> - [x] I built this PR locally for these architectures (if supported. mark crossbuilds):
> 
>   - [x] aarch64-musl
> 
>   - [x] armv7l
> 
>   - [x] armv6l-musl
> 
> 

seems that i was able to build this package locally through cross compile in `x86_64-musl`, `aarch64-musl`, `armv7l`, and `armv6l-musl`. going to test `i686`.

**Comment by paper42** [comment]
@ReneganRonin could you rebase this branch on top of master?

**Comment by ReneganRonin** [comment]
> seems that i was able to build this package locally through cross compile in `x86_64-musl`, `aarch64`, `aarch64-musl`, `armv7l`, and `armv6l-musl`. ~going to test `i686`.~
> `i686` built successfully.

done :)

**Comment by ReneganRonin** [comment]
> @ReneganRonin could you rebase this branch on top of master?

done :)

File: srcpkgs/noisetorch/INSTALL [hunk 7]
```diff
@@ -0,0 +1,5 @@
+case "${ACTION}" in
+post)
+	setcap 'CAP_SYS_RESOURCE=+ep' usr/bin/noisetorch
```
Comment (ericonr) [inline srcpkgs/noisetorch/INSTALL:3]: Needs `depends="libcap-progs"` in template

File: srcpkgs/noisetorch/template [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# Template file for 'noisetorch'
+pkgname=noisetorch
+version=0.11.3
+revision=1
+wrksrc=NoiseTorch-${version}
+build_style=go
+go_ldflags="-X main.version=${version} -X main.distribution=voidlinux"
```
Comment (ericonr) [inline srcpkgs/noisetorch/template:7]:
```suggestion
go_ldflags="-X main.version=${version} -X main.distribution=VoidLinux"
```

Nit

File: srcpkgs/noisetorch/template [hunk 9]
```diff
@@ -0,0 +1,26 @@
+# Template file for 'noisetorch'
+pkgname=noisetorch
```
Comment (ericonr) [inline srcpkgs/noisetorch/template:2]: Arguably the correct format would be `NoiseTorch`... But staying with this is valid.

File: srcpkgs/noisetorch/template [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+version=0.11.3
```
Comment (ericonr) [inline srcpkgs/noisetorch/template:3]: 0.11.4 is out

**Comment by uncomfyhalomacro** [comment]
This package is ready to merge and has no problems so far. Thanks!

File: srcpkgs/NoiseTorch/template [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# Template file for 'NoiseTorch'
+pkgname=NoiseTorch
+version=0.11.4
+revision=1
+wrksrc=${pkgname}-${version}
```
Comment (paper42) [inline srcpkgs/NoiseTorch/template:5]:
```suggestion
```

File: srcpkgs/NoiseTorch/template [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+build_style=go
+go_ldflags="-X main.version=${version} -X main.distribution=VoidLinux"
+go_import_path="."
+depends="libcap-progs"
+short_desc="Real-time microphone noise suppression on Linux"
+maintainer="Soc Virnyl S. Estela <renegan.ronin@gmail.com>"
+license="GPL-3.0-or-later"
+homepage="https://github.com/lawl/NoiseTorch"
+distfiles="${homepage}/archive/${version}.tar.gz"
```
Comment (paper42) [inline srcpkgs/NoiseTorch/template:14]:
```suggestion
distfiles="https://github.com/lawl/NoiseTorch/archive/${version}.tar.gz"
```

**Comment by ericonr** [comment]
Thanks for the PR! It worked great.
